adapters/from_sql.py

```diff
@@ -0,0 +1,194 @@
+    if most_recent_task:
+      # allows for continuation even if tasks have not begun/been generated yet
+      release.last_modified = to_timestamp(most_recent_task.last_modified)
+      release.last_active_task = most_recent_task.task_name
+    else:
+      release.last_modified = to_timestamp(item.execution_date)
+      release.last_active_task = ''
+    release_objects[release.release_id] = release
+
+  return release_objects
+
+
+def read_tasks(task_data):
+  """Reads the tasks, transforms them into objects.
+
+  Args:
+    task_data: a tuple of tuples with task info
+
+  Returns:
+    task_objects
+  """
+  task_named_tuple = namedtuple('Row', ['task_id', 'dag_id', 'execution_date', 'start_date', 'end_date', 'state'])
+  task_objects = []
+  for item in task_data:
+    task = Task()
+    logging.debug(item)
+    item = task_named_tuple._make(item)
+    task.task_name = item.task_id
+    task.add_dependency = None  # TODO(dommarques): figure out how to get this - dag info?
+    if item.start_date == None:
+      task.started = to_timestamp(item.execution_date)
+    else:
+      task.started = to_timestamp(item.start_date)
+    if item.state == None:
+      task.status = STATE_FROM_STRING.get('none')
+    else:
+      task.status = STATE_FROM_STRING.get(item.state)
+    if item.end_date == None:
+      task.last_modified = to_timestamp(item.execution_date)
+    else:
+      task.last_modified = to_timestamp(item.end_date)
+    task.log_url = 'https://youtu.be/dQw4w9WgXcQ'  # TODO(dommarques): figure out how to get the log in here
+    if item.state == None:
+      item.state == 'none'
+    else:
+      task.error = item.state
+    task_objects.append(task)
+  return task_objects
+
+
+def get_task_info(dag_id, execution_date, airflow_db):
+  """Gets task-related info to fill in missing release object info.
+
+  Args:
+    execution_date: a python datetime.datetime
+    airflow_db: the database connection object
+
+  Returns:
+    most_recent_task: the most recent active task
+    state: the state of the release based on the task states
+  """
+  task_query = to_sql_tasks(dag_id, execution_date)
+  task_data = airflow_db.query(task_query)
+  task_objects = read_tasks(task_data)
+  state = 0
+  task_ids = []
+  if task_objects:
+    most_recent_task = task_objects[len(task_objects) - 1]
+  else:
+    most_recent_task = None
+  for task in task_objects:
+    if task.status > state:
+      state = task.status
+    task_ids.append(task.task_name)
+  return task_ids, most_recent_task, state
+
+
+def read_xcom_vars(xcom_data):
+  """Reads the xcom data to get the dict of vars and the green build SHA."""
+  # occasionally, the xcom data doesn't exist yet. This is a workaround
+  xcom_named_tuple = namedtuple('Row', ['xcom_dict', 'green_sha'])
+  try:
+    xcom_data = xcom_named_tuple._make(xcom_data)
+    # The index 0 is here because the data returns in a tuple
+    xcom_dict = json.loads(xcom_data.xcom_dict[0])
```

Comment:
What does 0 represent ? Could you please define a global variable key ?

Reply:
The 0 is there only because the data returns in a tuple with only one entry, for example:
> xcom_dict = (some_data,)
Because of this, I didn't think a global variable key was necessary. I can make one if you wish though

Reply:
what is the other part of the tuple ? can we have get_xcom() method get rid of unused data ?

Reply:
There is no other part of the tuple, it is essentially a single var wrapped in a tuple. It would probably make sense to move the 0 index to get_xcom() though, so I'll do that.